lib/collection/src/collection_manager/optimizers/indexing_optimizer.rs

```diff
@@ -121,25 +121,71 @@ impl IndexingOptimizer {
                     return None; // Never optimize already optimized segment
                 }
 
-                // Apply indexing to plain segments which have grown too big
-                let are_all_vectors_indexed = segment_config.are_all_vectors_indexed();
-                let is_any_on_disk = segment_config.is_any_on_disk();
-
-                let big_for_mmap = vector_size
-                    >= self
-                        .thresholds_config
-                        .memmap_threshold
-                        .saturating_mul(BYTES_IN_KB);
-                let big_for_index = vector_size
-                    >= self
-                        .thresholds_config
-                        .indexing_threshold
-                        .saturating_mul(BYTES_IN_KB);
+                let indexing_threshold_kb = self
+                    .thresholds_config
+                    .indexing_threshold
+                    .saturating_mul(BYTES_IN_KB);
+                let mmap_threshold_kb = self
+                    .thresholds_config
+                    .memmap_threshold
+                    .saturating_mul(BYTES_IN_KB);
+                let mut require_optimization = false;
+
+                for (vector_name, vector_config) in self.collection_params.vectors.params_iter() {
+                    if let Some(vector_data) = segment_config.vector_data.get(vector_name) {
+                        let is_indexed = vector_data.index.is_indexed();
+                        let is_on_disk = vector_data.storage_type.is_on_disk();
+                        let storage_size = point_count * vector_data.size * VECTOR_ELEMENT_SIZE;
+
+                        let is_big_for_index = storage_size >= indexing_threshold_kb;
+                        let is_big_for_mmap = storage_size >= mmap_threshold_kb;
+
+                        let optimize_for_index = is_big_for_index && !is_indexed;
+                        let optimize_for_mmap = if let Some(on_disk_config) = vector_config.on_disk
+                        {
+                            on_disk_config && !is_on_disk
+                        } else {
+                            is_big_for_mmap && !is_on_disk
+                        };
+
+                        if optimize_for_index || optimize_for_mmap {
+                            require_optimization = true;
+                        }
+                    }
+                }
```

Comment:
If `require_optimization` is `true`, we don't need to look at the sparse vectors because they will be also handled by the optimizer.

WDYT about an early exit there

Reply:
Sure! Noticed this, but we implemented this quite hastily.

I'll make the change anyway :+1:

Reply:
Implemented in <9c7852438e7f494a3a53ecbf1a824d2b4457e88e>. We can improve it further after the release.